pkg/persistence/redis/redisclient.go

```diff
@@ -0,0 +1,160 @@
+package redis
+
+import (
+"errors"
+"fmt"
+"net/url"
+"strconv"
+"strings"
+"sync"
+"time"
+
+"github.com/FZambia/sentinel"
+"github.com/gomodule/redigo/redis"
+)
+
+var knownPool sync.Map
+var m sync.Mutex
+
+// RedisPoolParam ...
+type RedisPoolParam struct {
+	PoolMaxIdle     int
+	PoolMaxActive   int
+	PoolIdleTimeout time.Duration
+
+	DialConnectionTimeout time.Duration
+	DialReadTimeout       time.Duration
+	DialWriteTimeout      time.Duration
+}
+
+// GetRedisPool get a named redis pool
+// supported schema
+// redis://user:pass@redis_host:port/db
+// redis+sentinel://user:pass@redis_sentinel1:port1,redis_sentinel2:port2/monitor_name/db
+func GetRedisPool(name string, rawurl string, param *RedisPoolParam) (*redis.Pool, error) {
+	if p, ok := knownPool.Load(name); ok {
+		return p.(*redis.Pool), nil
+	}
+	m.Lock()
+	defer m.Unlock()
+	// load again in case multi threads
+	if p, ok := knownPool.Load(name); ok {
+		return p.(*redis.Pool), nil
+	}
+
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		return nil, fmt.Errorf("bad redis url: %s", err)
+	}
+
+	if param == nil {
+		param = &RedisPoolParam{
+			PoolMaxIdle:           0,
+			PoolMaxActive:         1,
+			PoolIdleTimeout:       time.Minute,
+			DialConnectionTimeout: time.Second,
+			DialReadTimeout:       time.Second,
+			DialWriteTimeout:      time.Second,
+		}
+	}
+
+	fmt.Println("get redis pool:", rawurl)
+	if u.Scheme == "redis" {
+		pool := &redis.Pool{
+			Dial: func() (redis.Conn, error) {
+				return redis.DialURL(rawurl)
+			},
+			TestOnBorrow: func(c redis.Conn, t time.Time) error {
+				_, err := c.Do("PING")
+				return err
+			},
+			MaxIdle:     param.PoolMaxIdle,
+			MaxActive:   param.PoolMaxActive,
+			IdleTimeout: param.PoolIdleTimeout,
+			Wait:        true,
+		}
+		return pool, nil
+	} else if u.Scheme == "redis+sentinel" {
+		return getSentinelPool(u, param, name)
+	} else if strings.HasPrefix(u.Scheme, "rediss") {
+		return nil, fmt.Errorf("bad redis url: not support secure redis")
+	} else {
+		return nil, fmt.Errorf("bad redis url: not support schema %s", u.Scheme)
+	}
+}
+
+// redis+sentinel://user:pass@redis_sentinel1:port1,redis_sentinel2:port2/monitor_name/db
+func getSentinelPool(u *url.URL, param *RedisPoolParam, name string) (*redis.Pool, error) {
+	ps := strings.Split(u.Path, "/")
```

Comment:
I believe that we need some unit tests for the URL parsing logic. Are you sure it's not going to split just after `redis+sentinel:/` ?

Reply:
We should also test that the right DialOptions are set based on the raw URL parse output

Reply:
u.Path is `redis_sentinel1:port1,redis_sentinel2:port2/monitor_name/db`

Reply:
This seems be problem, the u.Path I suppose to be `redis_sentinel1:port1,redis_sentinel2:port2/monitor_name/db`

Reply:
the u.Path is "/monitor_name/db"

Reply:
Yup, you're right